### ISSUE TITLE
ApolloServerPluginUsageReporting not working in Lambda environment

### ISSUE THREAD

**Issue by Danwakeem** [apollographql/apollo-server#5080]
## Description
My apollo studio graph does not appear to be receiving usage statictics from my AWS Lambda graphql service. I believe this could be an issue with `ApolloServerPluginUsageReporting` in the `apollo-server-core` package? 🤔 

I am using the following versions of apollo 👇 

```
"apollo-server-lambda": "^2.22.2"
```

And I have my server configured like this

```javascript
const server = new ApolloServer({
  schema: buildFederatedSchema([healthCheckSchema]),
  apollo: { key, graphVariant },
  plugins: [
    ApolloServerPluginUsageReporting({
      sendReportsImmediately: true,
      sendHeaders: { exceptNames: ['x-api-key'] },
      sendVariableValues: { all: true },
    }),
  ],
  introspection: true,
});
```

The important part being the `sendReportsImmediately` which is recommended when using this plugin in a serverless environment.

When I run the app locally using serverless-offline it writes the data to apollo studio just fine since my service doesn't shut down after every request 😉 

I am doing something similar in another app on an older version of apollo lambda (`"apollo-server-lambda": "^2.11.0"`) and it seems to be working fine there 🤔 But of course this could mean I am just doing the mapping from `engine` to `apollo` wrong so please let me know if that is the case 🤔 Also kind of interesting as I am not using the `sendReportsImmediately` flag here.

```javascript
const server = new ApolloServer({
  schema: buildFederatedSchema([healthCheckSchema]),
  engine: {
    apiKey: process.env.APOLLO_ENGINE_API_KEY,
    schemaTag: process.env.STAGE,
    privateHeaders: ['x-api-key'],
    sendVariableValues: {
      all: true,
    },
  },
  introspection: true,
});
```

### How to reproduce
I created a [bare bones serverless app repo](https://github.com/Danwakeem/apollo-serverless-demo) that can be deployed to AWS to reproduce the issue

### Screenshots
Here is the cloudwatch logs saying it is reporting something
<img width="1007" alt="Screen Shot 2021-04-02 at 9 28 34 PM" src="https://user-images.githubusercontent.com/1753824/113465891-fbc58680-93fc-11eb-9342-e1f5cb884588.png">

But when I following the link the reports never show up
<img width="1328" alt="Screen Shot 2021-04-02 at 9 51 29 PM" src="https://user-images.githubusercontent.com/1753824/113465974-aa69c700-93fd-11eb-90c9-92d7482f5321.png">

## TL;DR
- **Intended outcome:** I was expecting my usage result to be sent to Apollo studio before the lambda function completes execution.
- **Actual outcome:** Apollo studio never received the usage stats.
- **How to reproduce the issue:** I created a [bare bones serverless app repo](https://github.com/Danwakeem/apollo-serverless-demo) that can be deployed to AWS to reproduce the issue

> Sorry for the log post and if this is just me being dumb and not a real bug but I am not sure what I am doing wrong 😬

**Comment by glasser** [comment]
Thanks. I will try to look into this relatively soon, but is there any chance you can try to do some bisecting on the versions of apollo-server-lambda and apollo-server-core to see in what version this regression began? It sounds like it's OK in 2.11 and bad in 2.22.

I'd specifically be interested in hearing your results with:
- v2.18.2 (2.18.0 rewrote usage reporting entirely; 2.18.2 just fixes a few small bugs
- v2.20.0 (changed some stuff around lifecycle relating mostly to `stop()`)
- v2.21.2 (refactored the entire lambda package in a way that should have been backwards-compatible, but maybe not!)

Ideally try install both apollo-server-lambda and apollo-server-core at the same version (ie, install them both directly in your project's `dependencies` with an exact version "2.18.2" rather than range "^2.18.2").

**Comment by Danwakeem** [comment]
@glasser You know what, after doing some testing on different versions that you had suggested it actually looks like this _is working_ but the problem I was seeing was that some requests just never get sent to Apollo studio 🤔 

During my testing I would send 10 consecutive requests to my service using postman and I would only see 5 of the 10 requests show up in my studio operations dashboard.

If I send a single request this request would never show up in the dashboard.

I had tested with [quite a few version of apollo](https://github.com/Danwakeem/apollo-serverless-demo/branches/all) and they all actually had pretty similar results where only about half of the requests were logged to Apollo studio.

This seems to suggest that maybe the `sendReportsImmediately` flag isn't sending the reports immediately after a requests is received or I am misunderstanding something 🤔

**Comment by Danwakeem** [comment]
Assuming the `sendReportsImmediately` flag is supposed to send data before the lambda functions resolves 👇 

Looking through the plugin code it looks like this section of code might be the problem?

https://github.com/apollographql/apollo-server/blob/2a49de8ba8d5d85693e5656c9ff3b810c4eb9047/packages/apollo-server-core/src/plugin/usageReporting/plugin.ts#L647-L653

This is calling and not await results from `didEnd` which is also not awaiting the result from the `addTrace` function that appears to be looking at the `sendReportsImmediately` option and using `await` on `sendReportAndReportErrors` if set to `true`.

Since `didEnd` and `willSendResponse` are not awaiting `addTrace` that ends up calling the `sendReportAndReportErrors` this would result in the report not being sent until after the lambda function resolves which might explain why some graphql calls are not ever being sent to Apollo studio. 🤔 

When I add some console logs to my lambda function this appears to be the case (given the console logs are not lying to me about the order of execution 😉 )

<img width="898" alt="Screen Shot 2021-04-03 at 10 09 48 AM" src="https://user-images.githubusercontent.com/1753824/113482539-be94df00-9464-11eb-8c74-88d337fa7f0e.png">

**Comment by Danwakeem** [comment]
This is pretty ugly since I had to add two extra checks for `sendReportsImmediately` but just as a proof of concept I forked/created a branch with the change I suggested ☝️ and that appears to send every single request to Apollo studio when the `sendReportsImmediately` option is set 👇 

https://github.com/Danwakeem/apollo-server/blob/77afc8acc92c462119d1a709268ba25a4ddf06b2/packages/apollo-server-core/src/plugin/usageReporting/plugin.ts#L651-L661

Let me know what you think and I would be happy to help with a PR or any other troubleshooting ideas 😎

**Comment by glasser** [comment]
The code you linked to has been refactored as part of Apollo Server 4 and I hope it actually works. Specifically, `didEnd` is now called `operationFinished` and we do appear to await its call.
